File: python/examples/pathprobe.py
#!/usr/bin/env python3
"""pathprobe — invoke Client.parse_path and dump the encoded EPATH.
Mirrors c/examples/pathprobe.c CLI + output format.

SPDX-License-Identifier: MIT
"""
import sys

import bpclient
from bpclient.errors import BpEngine


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write(f"Usage: {sys.argv[0]} <text-path>\n")
        return 2
    text_path = sys.argv[1]

    try:
        client = bpclient.Client()
        client.open()
    except Exception:
        sys.stderr.write("client open failed\n")
        return 2
    try:
        client.open_session()
    except Exception:
        pass

    rc = 0
    parsed = None
    try:
        parsed = client.parse_path(text_path)
    except BpEngine as e:
        rc = e.code
    except Exception as e:
        rc = bpclient.err_code(e)

    encoded_len = len(parsed.encoded) if parsed else 0
    print(f"[pathprobe] text='{text_path}'  rc={rc}  "
          f"encoded_len={encoded_len}")
    if rc != 0 or parsed is None:
        client.close()
        return 1

    print(f"  out_class      = 0x{parsed.cip_class:04x}")
    print(f"  out_seg_flags  = 0x{parsed.segment_flags:02x}")
    print(f"  out_instance   = 0x{parsed.instance:08x}  ({parsed.instance})")
    print(f"  out_attr_flags = 0x{parsed.attr_flags:02x}")
    encoded_hex = " ".join(f"{b:02x}" for b in parsed.encoded)
    print(f"  encoded path   = {encoded_hex} ")
    client.close()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: c/examples/pathprobe.c
/*
 * pathprobe.c — invoke bp_client_parse_path and dump the encoded
 *               EPATH bytes it produces.  Useful when hand-building
 *               the route_path inside a routed Unconnected_Send
 *               (svc 0x52) carried in bp_client_message_send().cip_request.
 *
 * Usage:
 *   pathprobe "P:1,S:2"
 *   pathprobe "P:1,S:1"
 *   pathprobe "1,2,3"      (Rockwell-style — likely rejected)
 *
 * SPDX-License-Identifier: MIT
 */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bpclient.h"

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <text-path>\n", argv[0]);
        return 2;
    }
    bp_client_t *cl = NULL;
    if (bp_client_open(&cl) != BP_OK) {
        fprintf(stderr, "client open failed\n"); return 2;
    }
    bp_client_open_session(cl, NULL);

    bp_parsed_path_t parsed;
    int rc = bp_client_parse_path(cl, argv[1], &parsed);
    bp_client_close(cl);

    printf("[pathprobe] text='%s'  rc=%d  encoded_len=%u\n",
           argv[1], rc, parsed.encoded_size);
    if (rc != BP_OK) return 1;

    printf("  out_class      = 0x%04x\n", parsed.cip_class);
    printf("  out_seg_flags  = 0x%02x\n", parsed.segment_flags);
    printf("  out_instance   = 0x%08x  (%u)\n", parsed.instance, parsed.instance);
    printf("  out_attr_flags = 0x%02x\n", parsed.attr_flags);
    printf("  encoded path   = ");
    for (uint16_t i = 0; i < parsed.encoded_size; i++)
        printf("%02x ", parsed.encoded[i]);
    printf("\n");
    return 0;
}
